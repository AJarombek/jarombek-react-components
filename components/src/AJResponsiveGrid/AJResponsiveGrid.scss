/**
 * Stylesheet for the AJResponsiveGrid component.
 * Author: Andrew Jarombek
 * Date: 12/26/2019
 */

@import "../global";

.aj-responsive-grid {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  grid-gap: 20px;

  display: -ms-grid;
  -ms-grid-columns: 1fr 1fr 1fr;

  div {
    flex-grow: 1;
    flex-shrink: 0;
    flex-basis: auto;

    -ms-flex: 1 0 auto;
  }
}
