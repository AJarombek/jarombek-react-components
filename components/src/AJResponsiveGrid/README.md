### Overview

Reusable grid of cards which responds to different screen sizes.

**Use cases:**

```javascript
import { AJResponsiveGrid } from 'jarombek-react-components';

const CustomComponent = () => 
  <AJResponsiveGrid
    smallBreakpoint="300px"
    mediumBreakpoint="500px"
    largeBreakpoint="700px"
    items={[
      <AJTextCard
        title="Friday, February 21st"
        subtitle={<>Long Week</>}
        content={<>Excited to get a long sleep with my little friends Dotty & Lily after a long week.</>}
        action={f=>f}
        actionText="More Info"
      />
    ]} />
```

### Files

| Filename                   | Description                                                                     |
|----------------------------|---------------------------------------------------------------------------------|
| `AJResponsiveGrid.js`      | The reusable `AJResponsiveGrid` component.                                      |
| `index.js`                 | Index file which exports the `AJResponsiveGrid` component.                      |
| `styles.js`                | JSS styles for the `AJResponsiveGrid` component.                                |
