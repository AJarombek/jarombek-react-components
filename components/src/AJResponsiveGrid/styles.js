/**
 * Stylesheet for the AJResponsiveGrid component using JSS.
 * Author: Andrew Jarombek
 * Date: 12/26/2019
 */

export default {
  ajResponsiveGrid: ({small, medium, large}) => ({
    display: 'grid; display:-ms-grid',
    gridTemplateColumns: '1fr 1fr 1fr',
    gridGap: '20px',
    '-ms-grid-columns': '1fr 1fr 1fr',
    [`@media screen and (max-width: ${large}) and (min-width: ${medium})`]: {
      gridTemplateColumns: '1fr 1fr',
      gridGap: '20px',
      '-ms-grid-columns': '1fr 1fr'
    },
    [`@media screen and (max-width: ${medium}) and (min-width: ${small})`]: {
      gridTemplateColumns: '1fr',
      gridGap: '15px',
      '-ms-grid-columns': '1fr'
    },
    [`@media screen and (max-width: ${small})`]: {
      gridTemplateColumns: '1fr',
      gridGap: '10px',
      '-ms-grid-columns': '1fr'
    }
  })
};
