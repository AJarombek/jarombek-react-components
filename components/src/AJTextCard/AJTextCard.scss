@import "../global";

.aj-text-card {
  @include RobotoThin;
  display: inline-block;
  min-width: 300px;
  height: 500px;
  padding: 5px 10px;
  border-radius: 3px;
  box-shadow: 0 1px 3px rgba(170, 170, 170, 0.5);
}
