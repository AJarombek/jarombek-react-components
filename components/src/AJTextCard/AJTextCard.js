/**
 * Reusable generic card component which displays text.
 * @author Andrew Jarombek
 * @since 11/16/2019
 */

import React from 'react';
import PropTypes from 'prop-types';
import {createUseStyles} from 'react-jss';

import AJTextButton from '../AJTextButton/AJTextButton';
import styles from './styles';

const useStyles = createUseStyles(styles);

/**
 * Component representing a card that displays text and has an action button.
 * @param title Main title of the card.  Can be text or additional React elements.
 * @param subtitle Subtitle of the card.  Can be text or additional React elements.
 * @param content Text content of the card.  Can be text or additional React elements.
 * @param action Function that is called when the action button is clicked.
 * @param actionText Text of the action button in the footer of the card.  Can be text or
 * additional React elements.
 * @param actionDisabled Whether the action button is disabled from being clicked or not.
 * @return {*} React elements that make up a text card.
 */
const AJTextCard = ({title, subtitle, content, action, actionText, actionDisabled=false}) => {
  const classes = useStyles();

  return (
    <div className={classes.ajTextCard}>
      <div className="aj-text-card-header">
        <h2 className="aj-text-card-title">{title}</h2>
        <p className="aj-text-card-subtitle">{subtitle}</p>
      </div>
      <div className="aj-text-card-body">
        <p className="aj-text-card-content">{content}</p>
      </div>
      <div className="aj-text-card-footer">
        <AJTextButton onClick={action} disabled={actionDisabled}>{actionText}</AJTextButton>
      </div>
    </div>
  );
};

AJTextCard.propTypes = {
  title: PropTypes.node.isRequired,
  subtitle: PropTypes.node,
  content: PropTypes.node.isRequired,
  action: PropTypes.func.isRequired,
  actionText: PropTypes.node.isRequired,
  actionDisabled: PropTypes.bool
};

export default AJTextCard;
