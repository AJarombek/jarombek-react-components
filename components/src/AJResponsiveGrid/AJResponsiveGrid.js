/**
 * Reusable responsive grid component designed to fill an entire page.
 * @author Andrew Jarombek
 * @since 12/26/2019
 */

import React from 'react';
import PropTypes from 'prop-types';
import AJTextCard from '../AJTextCard/AJTextCard';
import {createUseStyles} from 'react-jss';
import styles from './styles';
import classnames from 'classnames';

const useStyles = createUseStyles(styles);

/**
 * Component representing a responsive grid of cards.  This grid is designed to take up an entire
 * page, although it can be used in other scenarios as well.
 * @param items A list of components which will be displayed in the grid.  These components should
 * be instances of {@link ../AJTextCard/AJTextCard}.
 * @param smallBreakpoint The smallest screen size breakpoint in the responsive grid.
 * @param mediumBreakpoint A medium sized screen breakpoint in the responsive grid.
 * @param largeBreakpoint The largest screen size breakpoint in the responsive grid.
 * @return {*} React elements that represent a responsive grid of cards.
 * @constructor
 */
const AJResponsiveGrid = ({items, smallBreakpoint='600px', mediumBreakpoint='900px',
                            largeBreakpoint='1200px'}) => {
  const breakpoints = {
    small: smallBreakpoint,
    medium: mediumBreakpoint,
    large: largeBreakpoint
  };

  const classes = useStyles(breakpoints);

  return (
    <div className={classnames('aj-responsive-grid', classes.ajResponsiveGrid)}>
      {items}
    </div>
  );
};

AJResponsiveGrid.propTypes = {
  items: PropTypes.arrayOf(PropTypes.object).isRequired,
  smallBreakpoint: PropTypes.string,
  mediumBreakpoint: PropTypes.string,
  largeBreakpoint: PropTypes.string
};

export default AJResponsiveGrid;
