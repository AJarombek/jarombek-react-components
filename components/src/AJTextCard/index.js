import AJTextCard from './AJTextCard';

export default AJTextCard;
