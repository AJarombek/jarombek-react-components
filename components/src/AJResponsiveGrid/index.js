import AJResponsiveGrid from './AJResponsiveGrid';

export default AJResponsiveGrid;
