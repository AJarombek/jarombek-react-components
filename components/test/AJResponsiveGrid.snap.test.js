import React from 'react';
import AJResponsiveGrid from '../src/AJResponsiveGrid/AJResponsiveGrid';
import AJTextCard from '../src/AJTextCard/AJTextCard';
import renderer from 'react-test-renderer';

/**
 * Snapshot test for the {@link ../../components/AJResponsiveGrid} component.
 * @author Andrew Jarombek
 * @since 12/28/2019
 */

it('renders correctly', () => {
  const tree = renderer.create(
    <AJResponsiveGrid
      smallBreakpoint="300px"
      mediumBreakpoint="500px"
      largeBreakpoint="700px"
      items={[
        <AJTextCard
          key="1"
          title="Snapshot Test 1"
          subtitle={<>subtitle</>}
          content={<>content</>}
          action={f=>f}
          actionText="More Info"
        />,
        <AJTextCard
          key="2"
          title="Snapshot Test 2"
          subtitle={<>subtitle</>}
          content={<>content</>}
          action={f=>f}
          actionText="More Info"
        />
      ]} />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
