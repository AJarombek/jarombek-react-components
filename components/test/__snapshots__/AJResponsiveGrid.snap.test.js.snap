// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="aj-responsive-grid ajResponsiveGrid-0-2-1 ajResponsiveGrid-d0-0-2-2"
>
  <div
    className="ajTextCard-0-2-3 aj-text-card"
  >
    <div
      className="aj-text-card-header"
    >
      <h2
        className="aj-text-card-title"
      >
        Snapshot Test 1
      </h2>
      <p
        className="aj-text-card-subtitle"
      >
        subtitle
      </p>
    </div>
    <div
      className="aj-text-card-body"
    >
      <p
        className="aj-text-card-content"
      >
        content
      </p>
    </div>
    <div
      className="aj-text-card-footer"
    >
      <div
        className="ajBaseButton-0-2-7 ajBaseButtonEnabled-0-2-8 aj-text-button ajTextButton-0-2-4 ajTextButtonEnabled-0-2-5"
        onClick={[Function]}
      >
        <button
          disabled={false}
          type="button"
        >
          More Info
        </button>
      </div>
    </div>
  </div>
  <div
    className="ajTextCard-0-2-3 aj-text-card"
  >
    <div
      className="aj-text-card-header"
    >
      <h2
        className="aj-text-card-title"
      >
        Snapshot Test 2
      </h2>
      <p
        className="aj-text-card-subtitle"
      >
        subtitle
      </p>
    </div>
    <div
      className="aj-text-card-body"
    >
      <p
        className="aj-text-card-content"
      >
        content
      </p>
    </div>
    <div
      className="aj-text-card-footer"
    >
      <div
        className="ajBaseButton-0-2-7 ajBaseButtonEnabled-0-2-8 aj-text-button ajTextButton-0-2-4 ajTextButtonEnabled-0-2-5"
        onClick={[Function]}
      >
        <button
          disabled={false}
          type="button"
        >
          More Info
        </button>
      </div>
    </div>
  </div>
</div>
`;
